crates/core/tedge_api/src/alarm.rs

```diff
@@ -1,58 +1,84 @@
-use std::convert::TryFrom;
-use std::convert::TryInto;
-use std::fmt;
-
+use crate::entity_store::EntityMetadata;
 use clock::Timestamp;
 use serde::Deserialize;
 use serde::Serialize;
-
 use serde_json::Value;
 use std::collections::HashMap;
-/// In-memory representation of ThinEdge JSON alarm.
-#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
+use std::convert::TryFrom;
+
+/// Internal representation of ThinEdge alarm model.
+#[derive(Debug, Eq, PartialEq)]
 pub struct ThinEdgeAlarm {
-    pub name: String,
-    pub severity: AlarmSeverity,
-    #[serde(flatten)]
-    pub data: Option<ThinEdgeAlarmData>,
-    pub source: Option<String>,
-}
+    /// Alarm type retrieved from the MQTT topic or payload.
+    /// The "type" given from the payload has priority to the one from the topic.
+    pub alarm_type: Option<String>,
 
-#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
-pub enum AlarmSeverity {
-    Critical,
-    Major,
-    Minor,
-    Warning,
+    /// Alarm source retrieved from the MQTT topic.
+    pub source: EntityMetadata,
+
+    /// All other alarm data retrieved from the MQTT payload.
+    /// None means the message is meant to clear an alarm.
+    pub data: Option<ThinEdgeAlarmData>,
 }
 
-/// In-memory representation of ThinEdge JSON alarm payload
+/// Internal representation of the JSON MQTT payload.
 #[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
 pub struct ThinEdgeAlarmData {
+    pub severity: Option<String>,
```

Comment:
It's thin-edge level. I thought the new scheme would allow full freedom for any fields. Rather, why do we need to have limited `severity` as Cumulocity?